pytorch_blade/pytorch_blade/compiler/jit/torch/shape_analysis_test.cpp

```diff
@@ -18,38 +18,116 @@
 namespace torch {
 namespace blade {
 
+void eraseInputShape(const std::shared_ptr<torch::jit::Graph>& graph) {
+  for (auto input : graph->inputs()) {
+    if (auto type = input->type()->cast<c10::TensorType>()) {
+      input->setType(type->dimensionedOnly());
+    }
+  }
+}
+
+#define FILE_CHECK(graph_str, s_pattern, dy_pattern)          \
+  auto g = std::make_shared<torch::jit::Graph>();             \
+  torch::jit::parseIR(graph_str, g.get());                    \
+  std::cout << g->toString() << std::endl;                    \
+  torch::blade::PropagateInputShapes(g);                      \
+  std::cout << g->toString() << std::endl;                    \
+  torch::jit::testing::FileCheck().check(s_pattern)->run(*g); \
+  eraseInputShape(g);                                         \
+  torch::blade::PropagateInputShapes(g);                      \
+  std::cout << g->toString() << std::endl;                    \
+  torch::jit::testing::FileCheck().check(dy_pattern)->run(*g);
+
 TEST(PropagateInputShapesTest, SimpleUnary) {
 #if PYTORCH_MAJOR_VERSOIN == 1 && PYTORCH_MINOR_VERSION >= 8
   const std::string graph_str = R"IR(
 graph(%p1 : Float(*, *, *, device=cuda:0)):
   %1 : Tensor = aten::relu(%p1)
   return (%1)
 )IR";
+  const std::string pattern = "Float(*, *, *) = aten::relu(%p1)";
 #elif PYTORCH_MAJOR_VERSION == 1 && PYTORCH_MINOR_VERSION > 6
   const std::string graph_str = R"IR(
 graph(%p1 : Float(*, *, *, device=cuda:0)):
   %1 : Tensor = aten::relu(%p1)
   return (%1)
 )IR";
+  const std::string pattern = "Float(*, *, *, device=cuda:0) = aten::relu(%p1)";
 #else
   const std::string graph_str = R"IR(
 graph(%p1 : Float(*, *, *)):
   %1 : Tensor = aten::relu(%p1)
   return (%1)
 )IR";
+  const std::string pattern = "Float(*, *, *) = aten::relu(%p1)";
 #endif
-  auto g = std::make_shared<torch::jit::Graph>();
-  torch::jit::parseIR(graph_str, g.get());
-  torch::blade::PropagateInputShapes(g);
-  std::cout << g->toString() << std::endl;
-  torch::jit::testing::FileCheck()
-      .check(
-#if PYTORCH_MAJOR_VERSION == 1 && PYTORCH_MINOR_VERSION <= 6
-          "Float(*, *, *) = aten::relu(%p1)")
-#else
-          "Float(*, *, *, device=cuda:0) = aten::relu(%p1)")
-#endif
-      ->run(*g);
+  FILE_CHECK(graph_str, pattern, pattern);
+}
+
+#if PYTORCH_MAJOR_VERSION == 1 && PYTORCH_MINOR_VERSION >= 12
```

Comment:
That's awesome!

Reply:
Shall we create a tool that takes a file as input? It will do shape propagation and file check on the input file. And the file contains the original graph without concrete shapes and the output graph pattern we want.

Reply:
that's a good idea.